### ISSUE TITLE
Sampling from different data set

### ISSUE THREAD

**Issue by ttdd11** [webdataset/webdataset#457]
First off - awesome repo - such a huge added value for all torch users.

In our particular use case, we consume data from multiple data sources of different sizes and train on multi-node tpu machines (~45M samples).

Currently, we combine all data sources together, shuffle, create the tars, and upload. In an effort to prevent one data source from dominating the training, we repeat the smaller data sets multiple times so that they are present in many shards and are seen frequently. This is of course very time consuming as if we change one smaller data set, we need to re-create all tars.

I'm wondering if it's possible to sample from multiple data sources, even if that means in the particular sources there could be repeats (to achieve the purpose above). 

Here is how we put together the data set: 

self.dataset = (wds.WebDataset(train_list,
              workersplitter=wds.split_by_worker,
              nodesplitter=wds.split_by_node,
              resampled=True,
              shardshuffle=True
              ).decode("pil").map(self.decode_sample).batched(train_size,partial=partial))

with a standard torch data loader. 

In this example, if train_list contains a list from all data sets (in the proportions we would like to train on), my understanding is that individual shrds will be assigned to each node, so some nodes will have data from some data sets, and others will have data from others, provided that we don't combine and shuffle ahead of time. Ideally, each minibatch would contain data from all data sets so it's not skewed depending on the data from each node. 

Long post for a simple question - is it possible to combine data from different sources and ensure that each node has a representative distribution of the data without combining and shuffling ahead of time? 

Thank you again for the awesome work and I'm happy to provide more information if this isn't clear.

**Comment by arijit-hub** [comment]
Hi,

I feel you can make use of `RandomMix`. 

You can create wds from different data sources and mix them based on the probability you want your model to see.

```python
ds1 = wds.WebDataset("first_source_tar_filepaths")
ds2 = wds.WebDataset("first_source_tar_filepaths")
...
dsn = ....
ds = wds.RandomMix([ds1, ds2, ..., dsn], probs=[0.1, 0.2,...0.1])
```

Hope this helps.

**Comment by ttdd11** [comment]
@arijit-hub thanks for the help - just have some questions on how to actually use this. So we create the random mix which samples from these tars, which is straight forward and thank you for the help there, then how do we get that into a torch dataloader?

Typically we use a process like this:

            wds.WebDataset(train_list,
                            #resampled=True,
                            workersplitter=wds.split_by_worker,
                            #cache_dir=cache_dir,
                            #cache_size=5e+9,
                            resampled=True,
                            shardshuffle=True
                        ).decode("pil").map(self.decode_sample).batched(train_size,partial=partial))

So when we create the mix, is the idea to set up each webdataset with the same inputs as above? How is the batching handled using randommix?

Thanks again.
